pallets/foreign-investments/src/impls/mod.rs

```diff
@@ -0,0 +1,1208 @@
+					// The minimum fulfillment must be everything
+					swap.amount,
+				)?;
+				ForeignInvestmentInfo::<T>::insert(
+					swap_order_id,
+					ForeignInvestmentInfoOf::<T> {
+						owner: who.clone(),
+						id: investment_id,
+						last_swap_reason: reason,
+					},
+				);
+			}
+			// Swap order either has not existed at all or was just cancelled
+			_ => {
+				let swap_order_id = T::TokenSwaps::place_order(
+					who.clone(),
+					swap.currency_in,
+					swap.currency_out,
+					swap.amount,
+					// The max accepted sell rate is independent of the asset type for now
+					T::DefaultTokenSellRate::get(),
+					// The minimum fulfillment must be everything
+					swap.amount,
+				)?;
+				TokenSwapOrderIds::<T>::insert(who, investment_id, swap_order_id);
+				ForeignInvestmentInfo::<T>::insert(
+					swap_order_id,
+					ForeignInvestmentInfoOf::<T> {
+						owner: who.clone(),
+						id: investment_id,
+						last_swap_reason: reason,
+					},
+				);
+			}
+		};
+		Ok(())
+	}
+
+	/// Determines the correct amount of a token swap based on the current
+	/// `InvestState` and `RedeemState` corresponding to the `TokenSwapOrderId`.
+	///
+	/// Returns a tuple of the total swap order amount as well as potentially
+	/// altered invest and redeem states. Any returning tuple element which is
+	/// `None`, reflects that no change is required for this element. Else, it
+	/// needs to be applied to the storage.
+	///
+	/// NOTE: Required since there exists at most one swap per `(AccountId,
+	/// InvestmentId)` pair whereas investments and redemptions can both mutate
+	/// orders. Assume, as a result of an `InvestState` transition, a token swap
+	/// order into pool currency is initialized. Then, as a result of a
+	/// `RedeemState` transition, a token swap order into foreign currency is
+	/// needed. This handler resolves the _merge conflict_ in situations where
+	/// the reason to create/update a swap order does not match the previous
+	/// reason.
+	///
+	/// * Is noop, if the the current reason equals the previous one.
+	/// * If both states are swapping into foreign currency, i.e. their invest
+	///   and redeem states include `ActiveSwapIntoForeignCurrency`, the states
+	///   stay the same. However the total order amount needs to be updated by
+	///   summing up both swap order amounts.
+	/// * If the `InvestState` includes swapping into pool currency, i.e.
+	///   `ActiveSwapIntoPoolCurrency`, whereas the `RedeemState` is swapping
+	///   into the opposite direction, i.e. `ActiveSwapIntoForeignCurrency`, we
+	///   need to resolve the delta between both swap order amounts and update
+	///   the states accordingly.
+	#[allow(clippy::type_complexity)]
+	fn handle_concurrent_swap_orders(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+	) -> Result<
+		(
+			Option<Swap<T::Balance, T::CurrencyId>>,
+			Option<InvestState<T>>,
+			Option<RedeemState<T::Balance, T::CurrencyId>>,
+			Option<TokenSwapReason>,
+		),
+		DispatchError,
```

Comment:
**MAYBE-CR**

Is it correct to assume that 
* `invest_swap.currency_in` = `pool_currency`
* `redeem_swap.currency_out` = `pool_currency`

Under this assumption: 
* `invest_swap.currency_in == redeem_swap.currency_out` &rarr; ALWAYS true 
* `invest_swap.currency_in == redeem_swap.currency_in` &rarr; NEVER true
    *The reason is that a user would need to redeem into the `pool_currency` and therefore will never have a swapping state.*
* `invest_swap.currency_out == redeem_swap.currency_out`  &rarr; NEVER true
    *The reason is that a user would receive a non-`pool_currency` for his redemption which is impossible*
* `invest_swap.currency_out == redeem_swap.currency_in` &rarr; The only case we need to match IMO

Reply:
> Is it correct to assume that
> 
> * `invest_swap.currency_in` = `pool_currency`
> * `redeem_swap.currency_out` = `pool_currency`

The `invest_swap.currency_in` can also be `foreign_currency`. The goal of this function is to handle the merge conflict if both the current `InvestmentState` as well as `RedemptionState` include an active swap. The handler is short-circuiting the cases in which at most one of these includes an active swap.

Reply:
So these are merging conflicts:
* Both swaps have the same direction: Swap amount must be sum of both active swaps and `TokenSwapReasons` set to `InvestmentAndRedemption` https://github.com/centrifuge/centrifuge-chain/blob/453a11124d9fec3765744bdf3c54b6452ef51111/pallets/foreign-investments/src/impls/mod.rs#L918
* Opposite direction with `invest_swap.amount` > `redeem_swap.amount`: "Fulfill" swap from `RedeemState` and decrease `invest_swap.amount` by that amount https://github.com/centrifuge/centrifuge-chain/blob/453a11124d9fec3765744bdf3c54b6452ef51111/pallets/foreign-investments/src/impls/mod.rs#L974
* Opposite direction with `invest_swap.amount` <= `redeem_swap.amount`: "Fulfill" swap from `InvestState` and decrease `redeem_swap.amount` by that amount https://github.com/centrifuge/centrifuge-chain/blob/453a11124d9fec3765744bdf3c54b6452ef51111/pallets/foreign-investments/src/impls/mod.rs#L985

Reply:
Ah. `invest_currency_in` can also be `foreign_currency` in cases where an order was decreased, right?

Reply:
Yes, exactly. But redemption swaps can only be from `pool` to `foreign`.